### PR TITLE
Support CanvasRenderingContext2D layers

### PR DESCRIPTION
**Pull request by shallawa**
#### 2b7e3e7edde7e861866efb5796cdcf6bedb513c1
<pre>
Support CanvasRenderingContext2D layers
<a href="https://bugs.webkit.org/show_bug.cgi?id=273923">https://bugs.webkit.org/show_bug.cgi?id=273923</a>
<a href="https://rdar.apple.com/127789082">rdar://127789082</a>

Reviewed by Kimmo Kinnunen.

This is a draft patch for the canvas layers API. The discussion of this API is
in <a href="https://github.com/whatwg/html/pull/9537.">https://github.com/whatwg/html/pull/9537.</a>

Unlike what is stated in the discussion above, this PR adds the filter to the
layer rendering states. This means like other members in layer rendering state,
the filter will be cleared out from the context state once beginLayer() is called.
It will restored to the context state once endLayer() is called. Then the restored
filter will be applied to the layer before compositing it to the canvas context.

* LayoutTests/fast/canvas/canvas-layer-alpha-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-layer-alpha-drawing.html: Added.
* LayoutTests/fast/canvas/canvas-layer-composite-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-layer-composite-drawing.html: Added.
* LayoutTests/fast/canvas/canvas-layer-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-layer-drawing.html: Added.
* LayoutTests/fast/canvas/canvas-layer-filter-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-layer-filter-drawing.html: Added.
* LayoutTests/fast/canvas/canvas-multiple-layer-filter-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-multiple-layer-filter-drawing.html: Added.
* LayoutTests/fast/canvas/canvas-multiple-nested-layer-filter-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-multiple-nested-layer-filter-drawing.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/CanvasFilterContextSwitcher.cpp: Copied from Source/WebCore/html/canvas/CanvasFilterTargetSwitcher.h.
(WebCore::CanvasFilterContextSwitcher::create):
(WebCore::CanvasFilterContextSwitcher::CanvasFilterContextSwitcher):
(WebCore::CanvasFilterContextSwitcher::~CanvasFilterContextSwitcher):
(WebCore::CanvasFilterContextSwitcher::expandedBounds const):
* Source/WebCore/html/canvas/CanvasFilterContextSwitcher.h: Copied from Source/WebCore/html/canvas/CanvasFilterTargetSwitcher.h.
* Source/WebCore/html/canvas/CanvasLayerContextSwitcher.cpp: Renamed from Source/WebCore/html/canvas/CanvasFilterTargetSwitcher.cpp.
(WebCore::CanvasLayerContextSwitcher::create):
(WebCore::CanvasLayerContextSwitcher::CanvasLayerContextSwitcher):
(WebCore::CanvasLayerContextSwitcher::~CanvasLayerContextSwitcher):
(WebCore::CanvasLayerContextSwitcher::drawingContext const):
(WebCore::CanvasLayerContextSwitcher::outsets const):
* Source/WebCore/html/canvas/CanvasLayerContextSwitcher.h: Renamed from Source/WebCore/html/canvas/CanvasFilterTargetSwitcher.h.
* Source/WebCore/html/canvas/CanvasLayers.idl: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
(WebCore::CanvasRenderingContext2D::createFilter const):
* Source/WebCore/html/canvas/CanvasRenderingContext2D.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::beginLayer):
(WebCore::CanvasRenderingContext2DBase::endLayer):
(WebCore::CanvasRenderingContext2DBase::fillInternal):
(WebCore::CanvasRenderingContext2DBase::strokeInternal):
(WebCore::CanvasRenderingContext2DBase::fillRect):
(WebCore::CanvasRenderingContext2DBase::strokeRect):
(WebCore::CanvasRenderingContext2DBase::drawImage):
(WebCore::CanvasRenderingContext2DBase::drawingContext const):
(WebCore::CanvasRenderingContext2DBase::effectiveDrawingContext const):
(WebCore::CanvasRenderingContext2DBase::drawTextUnchecked):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::createFilter const):
(WebCore::CanvasRenderingContext2DBase::setFilterTargetSwitcher): Deleted.
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.idl:
* Source/WebCore/platform/graphics/GraphicsContextSwitcher.cpp: Renamed from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp.
(WebCore::GraphicsContextSwitcher::create):
(WebCore::GraphicsContextSwitcher::GraphicsContextSwitcher):
* Source/WebCore/platform/graphics/GraphicsContextSwitcher.h: Renamed from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.h.
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::filteredNativeImage):
* Source/WebCore/platform/graphics/ImageBufferContextSwitcher.cpp: Renamed from Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.cpp.
(WebCore::ImageBufferContextSwitcher::ImageBufferContextSwitcher):
(WebCore::ImageBufferContextSwitcher::drawingContext const):
(WebCore::ImageBufferContextSwitcher::beginClipAndDrawSourceImage):
(WebCore::ImageBufferContextSwitcher::endClipAndDrawSourceImage):
(WebCore::ImageBufferContextSwitcher::endDrawSourceImage):
* Source/WebCore/platform/graphics/ImageBufferContextSwitcher.h: Renamed from Source/WebCore/platform/graphics/filters/FilterImageTargetSwitcher.h.
* Source/WebCore/platform/graphics/TransparencyLayerContextSwitcher.cpp: Renamed from Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp.
(WebCore::TransparencyLayerContextSwitcher::TransparencyLayerContextSwitcher):
(WebCore::TransparencyLayerContextSwitcher::beginClipAndDrawSourceImage):
(WebCore::TransparencyLayerContextSwitcher::beginDrawSourceImage):
(WebCore::TransparencyLayerContextSwitcher::endDrawSourceImage):
* Source/WebCore/platform/graphics/TransparencyLayerContextSwitcher.h: Renamed from Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.h.
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.cpp:
(WebCore::LegacyRenderSVGResourceFilter::applyResource):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceFilter.h:

Canonical link: <a href="https://commits.webkit.org/280148@main">https://commits.webkit.org/280148@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f6b50c1cc70656c0f0339d9c22e42e3f740baa80

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55835 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35159 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8303 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58831 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6267 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42780 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6463 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44964 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4326 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57864 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33056 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48132 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26097 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29840 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5460 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4410 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48911 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5731 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60420 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55071 "Built successfully and passed tests") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5858 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52396 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48201 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51892 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12379 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/76832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30988 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/76832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32072 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33154 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31820 "Built successfully") | | |

### CODE REVIEW

**Review by karlcow — COMMENTED** [review]
I wonder if these are already covered by tests in https://wpt.fyi/results/html/canvas

```
* LayoutTests/fast/canvas/canvas-layer-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-layer-drawing.html: Added.
* LayoutTests/fast/canvas/canvas-layer-filter-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-layer-filter-drawing.html: Added.
* LayoutTests/fast/canvas/canvas-multiple-layer-filter-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-multiple-layer-filter-drawing.html: Added.
* LayoutTests/fast/canvas/canvas-multiple-nested-layer-filter-drawing-expected.html: Added.
* LayoutTests/fast/canvas/canvas-multiple-nested-layer-filter-drawing.html: Added.
```

I just tried to import `html/canvas` and indeed there are a lot of outdated stuff in WebKit repo, that would deserve an import. 


```
Import complete
IMPORTED 2423 TOTAL TESTS
Imported 358 reftests
Imported 2061 JS tests
Imported 4 Crash tests
Imported 4 pixel/manual tests
```

**Comment by karlcow** [comment]
> I just tried to import `html/canvas` and indeed there are a lot of outdated stuff in WebKit repo, that would deserve an import.

I opened https://bugs.webkit.org/show_bug.cgi?id=274304
I will see if I can straighten up these.

File: Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.idl [hunk 1]
```diff
@@ -40,6 +40,7 @@
 };
 
 OffscreenCanvasRenderingContext2D includes CanvasState;
+OffscreenCanvasRenderingContext2D includes CanvasLayers;
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.idl:43]: Does the `createFilter` work for offscreen canvas? ...2DBase has an empty implementation?

File: Source/WebCore/html/canvas/CanvasFilterContextSwitcher.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace WebCore {
+
+std::unique_ptr<CanvasFilterContextSwitcher> CanvasFilterContextSwitcher::create(CanvasRenderingContext2DBase& context, Function<FloatRect()>&& boundsProvider)
+{
+    if (context.state().filterOperations.isEmpty())
+        return nullptr;
+
+    FloatRect bounds;
+    auto filter = context.createFilter([&]() {
```
Comment (kkinnunen-apple) [inline Source/WebCore/html/canvas/CanvasFilterContextSwitcher.cpp]: The createFilter  API taking const Function& is a bug. by changing it to Function&& you should be able to just pass WTFMove(boundsProvider)

**Review by kkinnunen-apple — COMMENTED** [review]
FWIW the switcher concept is confusing to me. The source of it seems to be that we hold the filter as a top level concept. The stack becomes:
 - ImageBuffer
   - Switcher
     - Filter
       - ImageBuffer     
 
IMO the layers could be just stack of ImageBuffers:
 - ImageBuffer
   - ImageBuffer
     - ImageBuffer

With the filters being set state inside the ImageBuffers.

File: Source/WebCore/html/canvas/CanvasFilterContextSwitcher.cpp (hunk 2 above)
Comment (shallawa) [reply Source/WebCore/html/canvas/CanvasFilterContextSwitcher.cpp]: I changed it to be `'const FloatRect&` instead and I mode the check `if (state().filterOperations.isEmpty())` to the caller side.

**Comment by shallawa** [comment]
> FWIW the switcher concept is confusing to me. The source of it seems to be that we hold the filter as a top level concept. The stack becomes:
> 
> * ImageBuffer
>   
>   * Switcher
>     
>     * Filter
>       
>       * ImageBuffer
> 
> IMO the layers could be just stack of ImageBuffers:
> 
> * ImageBuffer
>   
>   * ImageBuffer
>     
>     * ImageBuffer
> 
> With the filters being set state inside the ImageBuffers.

The only problem with this is CG filters do not require ImageBuffers. The Switcher in this case is mainly a state tracker to being and end a transparency layer.

**Comment by kkinnunen-apple** [comment]
> The only problem with this is CG filters do not require ImageBuffers. The Switcher in this case is mainly a state tracker to being and end a transparency layer.

But in those cases I'd imagine save() and restore() be enough? Anyhow, not specific to this, it's already in.

**Review by kkinnunen-apple — APPROVED** [review]
IIUC, this is just a commit to get the work started? And further test failures will be fixed in further commits?
If I understand the spec, for example the exceptions are missing, at least on putImageData while filter is on, etc?
And OffscreenCanvas doesn't work by intention, yet? Maybe if OffscreenCanvas doesn't work, the .idl change could be removed..

**Pull request merged**